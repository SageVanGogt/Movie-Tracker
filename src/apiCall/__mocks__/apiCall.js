import mockData from './../../mockData/mockData';
const fetchRecentFilms = jest.fn().mockImplementation(() => Promise.resolve(mockData))
const addUserFetch = jest.fn().mockImplementation(() => Promise.resolve({
    name: "",
    email: "great",
    password: ""
}))

export {fetchRecentFilms, addUserFetch};

