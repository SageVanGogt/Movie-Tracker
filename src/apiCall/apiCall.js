import key from './apiKey';

const fetchRecentFilms = async () => {
  const url = `https://api.themoviedb.org/3/discover/movie?api_key=${key}&primary_release_date.gte=2014-09-15&primary_release_date.lte=2014-10-22`;
  try{
    const response = await fetch(url);
    const filmData = await response.json();
    return filmData;
  } catch(err){
    const error = 'Failed to fetch data'
    throw error;
  }
}

const addUserFetch = async (user) => {
  const url = "http://localhost:3000/api/users/new"
  try {
    const response = await fetch(url, {
      method: "POST",
      body: JSON.stringify(user),
      headers: {
        Accept: 'application/json',
        "Content-Type": "application/json"}
    })
    const userData = await response.json();
    return userData;
  } catch(err){
    const error = 'Failed to fetch data'
    throw error;
  }
}

const fetchUser = async (user) => {
  const email = user.email.toLowerCase();
  const  userBody = {
    method: 'POST',
    body: JSON.stringify({ email, password: user.password }),
    headers: {"Content-Type": "application/json"}
  }
  const url = 'http://localhost:3000/api/users';
  try {
    const response = await fetch(url, userBody)
    const userData = await response.json();
    return userData;
  } catch (err) {
    const error = 'Failed to fetch data'
    throw error;
  }
}

const postFavoriteToDb = async (favoriteInfo, user) => {
  const url = "http://localhost:3000/api/users/favorites/new";
  const postData = {
    method: 'POST',
    body: JSON.stringify({user_id: user.user_id, ...favoriteInfo}),
    headers: {
      Accept: 'application/json',
      "Content-Type": "application/json"
    }
  }
  try {
    await fetch(url, postData);
  } catch(err) {
    const error = "Failed to post favorite";
    throw error;
  }
}

const getUserFavorites = async (userId) => {
  const url = `http://localhost:3000/api/users/${userId}/favorites`;
  try{
  const response = await fetch(url);
  const favorites = await response.json();
  return favorites;
  } catch(err) {
    const error = "Failed to get favorites";
    throw error;
  }
}

const removeFavorite = async (favorite) => {
  debugger
  const url = `http://localhost:3000/api/users/${favorite.user_id}/favorites/${favorite.movie_id}`
  try{
    const response = await fetch(url, {
      method: 'DELETE',
      headers: {
        "Content-Type": "application/json",
        },
      body: JSON.stringify(favorite)

    })
    const data = await response.json();
  } catch(err) {
    const error = "Failed to remove favorite";
    throw error;
  }
}


export {
  fetchRecentFilms, 
  addUserFetch, 
  fetchUser,
  postFavoriteToDb,
  getUserFavorites,
  removeFavorite
};