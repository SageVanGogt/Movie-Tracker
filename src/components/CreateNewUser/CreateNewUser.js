import React, { Component } from "react";

class CreateNewUser extends Component {
  constructor(props) {
    super(props);

    this.state = {
      name: "",
      email: "",
      password: ""
    };
  }

  handleChange = (event) => {
    const {name, value} = event.target
    this.setState({
        [name]: value
    })
  };

  handleSubmit = (event) => {
    event.preventDefault();

  };

  render() {
    return (
      <form type="submit" onChange={this.handleSubmit}>
        <input
          type="text"
          value={this.state.name}
          name="name"
          onChange={this.handleChange}
        />
        <input
          type="text"
          value={this.state.email}
          name="email"
          onChange={this.handleChange}
        />
        <input
          type="text"
          value={this.state.password}
          name="password"
          onChange={this.handleChange}
        />
      </form>
    );
  }
}

export default CreateNewUser