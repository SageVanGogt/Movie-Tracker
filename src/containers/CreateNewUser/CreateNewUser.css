.new-user-form {
  display: inline-block;
  margin: 20px;
  box-sizing: border-box;
  height: 250px;
}

.border {
  border-right: 2px soild #FFF; 
}

.form-field {
  display: block;
  width: 350px;
  height: 30px;
  margin: 15px;
  padding-left: 10px;
}

.submit-button {
  display: block;
  width: 350px;
  background-color: #a108ff;
  color: #FFF;
  height: 30px;
  margin: 15px;
  box-sizing: content-box;
  border: none;
  cursor: pointer;
}

input[type="submit"]:hover {
  background-color: #b748fc;
}

.login-header {
  color: #FFF;
  font-size: 1.3rem;
  font-weight: 400;
  margin: 15px;

}