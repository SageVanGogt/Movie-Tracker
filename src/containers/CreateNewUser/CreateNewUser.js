import React, { Component } from "react";
import { addUserFetch, validateEmail } from "../../apiCall/apiCall";
import { connect } from 'react-redux';
import { updateStoreUser, addError } from './../../actions/index';
import PropTypes from 'prop-types';
import './CreateNewUser.css';

export class CreateNewUser extends Component {
  constructor(props) {
    super(props);

    this.state = {
      name: "",
      email: "",
      password: ""
    };
  }

  handleChange = (event) => {
    const {name, value} = event.target;
    this.setState({
      [name]: value
    });
  };

  checkValidEmail = async (event) => {
    event.preventDefault();    
    const actual = await validateEmail(this.state.email);
    if (actual.isValid) {
      this.handleSubmit();
    } else {
      this.props.handleError('That email is not valid') ;
    }
  };

  handleSubmit = async () => {
    const response = await addUserFetch(this.state);
    if (response.id) {
      this.props.handleSignup({id: response.id, name: this.state.name});
    } else if (response.error) {
      this.props.handleError("Email already in use");
    }
  };

  render() {
    return (
      <form 
        type="submit"
        className="new-user-form border" 
        onSubmit={this.checkValidEmail}
      >
        <h2 className="login-header">SIGN UP</h2>
        <input
          className = "form-field"
          type="text"
          value={this.state.name}
          name="name"
          onChange={this.handleChange}
          placeholder="name"          
        />
        <input
          className="form-field"
          type="text"
          value={this.state.email}
          name="email"
          onChange={this.handleChange}
          placeholder="email"          
        />
        <input
          className="form-field"
          type="text"
          value={this.state.password}
          name="password"
          onChange={this.handleChange}
          placeholder="password"          
        />
        <input
          className = "submit-button"
          type = "submit"
        />
      </form>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  handleSignup:(user) => dispatch(updateStoreUser(user)),
  handleError: (error) => dispatch(addError(error))    
});

CreateNewUser.propTypes = {
  handleSignup: PropTypes.func,
<<<<<<< HEAD
  user: PropTypes.object
};
=======
  user: PropTypes.object,
  error: PropTypes.string,
  handleError: PropTypes.func
}
>>>>>>> Add proptypes and fixed api test

export default connect(null, mapDispatchToProps)(CreateNewUser);