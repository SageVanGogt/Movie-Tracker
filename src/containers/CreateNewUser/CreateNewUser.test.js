import { CreateNewUser, mapDispatchToProps } from "./CreateNewUser";
import React from "react";
import { shallow } from "enzyme";
import { addUserFetch, validateEmail } from "./../../apiCall/apiCall";

jest.mock('./../../apiCall/apiCall');

describe("CreateNewUser", () => {
  let wrapper;
  let mockHandleSignup;
  let mockHandleError;

  beforeEach(() => {
    mockHandleSignup = jest.fn();
    mockHandleError = jest.fn();
    wrapper = shallow(<CreateNewUser 
      handleSignup={mockHandleSignup}
      handleError={mockHandleError}
    />);
  });

  it("has a default state", () => {
    let expected = {
      name: "",
      email: "",
      password: ""
    };

    expect(wrapper.state()).toEqual(expected);
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('handleChange', () => {
    it("should update state on change", () => {
      let mockEvent = {
        target: { name: "email", value: "great" }
      };
      let expected = {
        name: "",
        email: "great",
        password: ""
      };
  
      wrapper.instance().handleChange(mockEvent);
  
      expect(wrapper.state()).toEqual(expected);
    });
  });

  describe('handleSubmit', () => {
    it('should calls addUserFetch callback after adding user', async () => {
      Promise.resolve(wrapper.instance().handleSubmit());
      
      expect(addUserFetch).toHaveBeenCalledWith(wrapper.state());  
    });
  
    it('should call handleSignup with the correct params', async () => {
      const mockUser = {
        "id": 1,
        "name": ""
      };
  
      await wrapper.instance().handleSubmit();
  
      expect(mockHandleSignup).toHaveBeenCalledWith(mockUser);
    });
  });

  describe('mapDispatchtoProps', () => {
    it('should be called with the correct params', () => {
      const mockDispatch = jest.fn();
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = {
        type: "ADD_USER",
        user_id: 1,
        name: "doc"  
      };

      const mockUser = {
        id: 1,
        name: 'doc'
      };

      mappedProps.handleSignup(mockUser);

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });

    it('should call dispatch on handleError with the correct params', () => {
      const mockDispatch = jest.fn();
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = {
        type: "ADD_ERROR",
        message: "Email already in use"
      };
      const mockError = "Email already in use";

      mappedProps.handleError(mockError);

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });
  });

  describe('checkValidEmail', () => {
    let mockEvent;
    let mockState;
    beforeEach(() => {
      mockEvent = {
        preventDefault: jest.fn()
      };
      mockState = 'thurmanvogt@gmail.com';
      wrapper.setState({email: mockState});
    });

    it('should call validateEmail with the correct props', () => {
      wrapper.instance().checkValidEmail(mockEvent);      
      expect(validateEmail).toHaveBeenCalledWith(mockState);
    });

    it.skip('should call handleSubmit if actual.isValid is true', () => {
      const spy = jest.spyOn(CreateNewUser, 'handleSubmit');
      wrapper.instance().checkValidEmail(mockEvent);
      expect(spy).toHaveBeenCalled();
    });
  });
});
