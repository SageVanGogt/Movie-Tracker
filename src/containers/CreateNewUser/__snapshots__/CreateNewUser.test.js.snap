// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNewUser should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CreateNewUser
    handleSignup={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2
          className="login-header"
>
          SIGN UP
</h2>,
        <input
          className="form-field"
          name="name"
          onChange={[Function]}
          placeholder="name"
          type="text"
          value=""
/>,
        <input
          className="form-field"
          name="email"
          onChange={[Function]}
          placeholder="email"
          type="text"
          value=""
/>,
        <input
          className="form-field"
          name="password"
          onChange={[Function]}
          placeholder="password"
          type="text"
          value=""
/>,
        <input
          className="submit-button"
          type="submit"
/>,
      ],
      "className": "new-user-form border",
      "onSubmit": [Function],
      "type": "submit",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "SIGN UP",
          "className": "login-header",
        },
        "ref": null,
        "rendered": "SIGN UP",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "form-field",
          "name": "name",
          "onChange": [Function],
          "placeholder": "name",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "form-field",
          "name": "email",
          "onChange": [Function],
          "placeholder": "email",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "form-field",
          "name": "password",
          "onChange": [Function],
          "placeholder": "password",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "submit-button",
          "type": "submit",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2
            className="login-header"
>
            SIGN UP
</h2>,
          <input
            className="form-field"
            name="name"
            onChange={[Function]}
            placeholder="name"
            type="text"
            value=""
/>,
          <input
            className="form-field"
            name="email"
            onChange={[Function]}
            placeholder="email"
            type="text"
            value=""
/>,
          <input
            className="form-field"
            name="password"
            onChange={[Function]}
            placeholder="password"
            type="text"
            value=""
/>,
          <input
            className="submit-button"
            type="submit"
/>,
        ],
        "className": "new-user-form border",
        "onSubmit": [Function],
        "type": "submit",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "SIGN UP",
            "className": "login-header",
          },
          "ref": null,
          "rendered": "SIGN UP",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "form-field",
            "name": "name",
            "onChange": [Function],
            "placeholder": "name",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "form-field",
            "name": "email",
            "onChange": [Function],
            "placeholder": "email",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "form-field",
            "name": "password",
            "onChange": [Function],
            "placeholder": "password",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "submit-button",
            "type": "submit",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
